File: C++/leetcode707.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class MyLinkedList {
public:
    //定义结点结构
    struct LinkedNode{
        int val;
        LinkedNode* next;
        LinkedNode(int val):val(val),next(nullptr){}
    };
    LinkedNode* _dummyhead;
    //使用一个变量来记录链表长度
    int _size;


    MyLinkedList() {
        _dummyhead = new LinkedNode(0);
        _size = 0;
    }
    
    int get(int index) {
        //注
        if(index>=_size)
            return -1;
        LinkedNode* cur = _dummyhead;
        int i=0;
        while(i<=index){
            cur = cur->next;
            i++;
        }
        return cur->val;
    }
    
    void addAtHead(int val) {
        LinkedNode* p = new LinkedNode(val);
        p->next = _dummyhead->next;
        _dummyhead->next = p;
        _size++;
    }
    
    void addAtTail(int val) {
        LinkedNode* p = new LinkedNode(val);
        LinkedNode* cur = _dummyhead;
        while(cur->next!=NULL){
            cur = cur->next;
        }
        cur->next = p;
        _size++;
    }
    
    void addAtIndex(int index, int val) {
        //注
        if(index>_size)
            return;
        LinkedNode* cur = _dummyhead;
        while(index--){
            cur = cur->next;
        }
        LinkedNode* p = new LinkedNode(val);
        p->next = cur->next;
        cur->next = p;
        _size++;
    }
    
    void deleteAtIndex(int index) {
        //注
        if(index>=_size)
            return;


        LinkedNode* cur = _dummyhead;
        // 注
        while(index--){
            cur = cur->next;
        }
        LinkedNode* temp = cur->next;
        cur->next = cur->next->next;
        delete temp;

        //注
        _size--;
    }
};

File: python/leetcode707.py
class ListNode:
    def __init__(self,val = 0,next = None) -> None:
        self.val = val
        self.next = next


class MyLinkedList:

# 使用虚拟头结点，一个变量记录总长度

    def __init__(self):
        self._dummyHead = ListNode()
        self._size = 0

    def get(self, index: int) -> int:
        if index >=self._size:
            return -1
        


        cur = self._dummyHead.next
        for i in range(index):
            cur = cur.next

        return cur.val


    def addAtHead(self, val: int) -> None:
        newnode = ListNode(val=val)
        newnode.next = self._dummyHead.next
        self._dummyHead.next = newnode
        self._size +=1



    def addAtTail(self, val: int) -> None:
        cur = self._dummyHead
        while cur.next != None:
            cur = cur.next
        
        newnode = ListNode(val=val)
        cur.next = newnode
        self._size +=1



    def addAtIndex(self, index: int, val: int) -> None:
        if index >self._size:
            return
        cur = self._dummyHead
        while index:
            cur = cur.next
            index -= 1
        newnode = ListNode(val=val)
        newnode.next = cur.next
        cur.next = newnode
        self._size +=1


    def deleteAtIndex(self, index: int) -> None:
        if index >= self._size:
            return
        cur = self._dummyHead
        while index :
            cur = cur.next
            index -= 1

        cur.next = cur.next.next
        self._size -= 1